src/Shared/TryParseMethodCache.cs

```diff
@@ -4,57 +4,90 @@
+        private static readonly ConcurrentDictionary<Type, Func<Expression, MethodCallExpression>?> MethodCallCache = new();
+        internal static readonly ParameterExpression TempSourceStringExpr = Expression.Variable(typeof(string), "tempSourceString");
 
         public static bool HasTryParseMethod(ParameterInfo parameter)
         {
             var nonNullableParameterType = Nullable.GetUnderlyingType(parameter.ParameterType) ?? parameter.ParameterType;
             return FindTryParseMethod(nonNullableParameterType) is not null;
         }
 
-        // TODO: Use InvariantCulture where possible? Or is CurrentCulture fine because it's more flexible?
-        public static MethodInfo? FindTryParseMethod(Type type)
+        public static Func<Expression, MethodCallExpression>? FindTryParseMethod(Type type)
         {
-            static MethodInfo? Finder(Type type)
+            static Func<Expression, MethodCallExpression>? Finder(Type type)
             {
+                MethodInfo? methodInfo;
+
                 if (type.IsEnum)
                 {
-                    return EnumTryParseMethod.MakeGenericMethod(type);
+                    methodInfo = EnumTryParseMethod.MakeGenericMethod(type);
+                    if (methodInfo != null)
+                    {
+                        return (expression) => Expression.Call(methodInfo!, TempSourceStringExpr, expression);
+                    }
+
+                    return null;
                 }
 
-                var staticMethods = type.GetMethods(BindingFlags.Public | BindingFlags.Static);
+                if (TryGetDateTimeTryParseMethod(type, out methodInfo))
+                {
+                    return (expression) => Expression.Call(
+                        methodInfo!,
+                        TempSourceStringExpr,
+                        Expression.Constant(CultureInfo.InvariantCulture),
+                        Expression.Constant(DateTimeStyles.None),
+                        expression);
+                }
 
-                foreach (var method in staticMethods)
+                if (TryGetNumberStylesTryGetMethod(type, out methodInfo, out var numberStyle))
                 {
-                    if (method.Name != "TryParse" || method.ReturnType != typeof(bool))
-                    {
-                        continue;
-                    }
+                    return (expression) => Expression.Call(
+                        methodInfo!,
+                        TempSourceStringExpr,
+                        Expression.Constant(numberStyle),
+                        Expression.Constant(CultureInfo.InvariantCulture),
+                        expression);
+                }
 
-                    var tryParseParameters = method.GetParameters();
+                methodInfo = type.GetMethod("TryParse", BindingFlags.Public | BindingFlags.Static, new[] { typeof(string), typeof(IFormatProvider), type.MakeByRefType() });
```

Comment:
Can we add a test with a custom type that implements this signature and verify it actually gets called with `CultureInfo.InvariantCulture`?

Reply:
Sure ! 🙂